controllers/managedfusiondeployment_controller.go

```diff
@@ -0,0 +1,752 @@
+		if !utils.Contains(r.managedFusionSecret.GetFinalizers(), managedFusionFinalizer) {
+			r.Log.V(-1).Info("finalizer missing on the managed-fusion-agent-config secret resource, adding...")
+			r.managedFusionSecret.SetFinalizers(append(r.managedFusionSecret.GetFinalizers(), managedFusionFinalizer))
+			if err := r.update(r.managedFusionSecret); err != nil {
+				return ctrl.Result{}, fmt.Errorf("failed to update managed-fusion-agent-config secret with finalizer: %v", err)
+			}
+		}
+
+		// Check if the structure of managedFusionSecret is valid or not
+		managedFusionSecretData := r.managedFusionSecret.Data
+		managedFusionDeploymentPagerDutyConfig, exist := managedFusionSecretData["pager_duty_config"]
+		if !exist {
+			return ctrl.Result{}, fmt.Errorf("managed-fusion-agent-config secret does not contain pager_duty_config entry")
+		}
+		if err := yaml.Unmarshal(managedFusionDeploymentPagerDutyConfig, r.pagerDutyConfigData); err != nil {
+			return ctrl.Result{}, err
+		}
+		managedFusionDeploymentSMTPConfig, exist := managedFusionSecretData["smtp_config"]
+		if !exist {
+			return ctrl.Result{}, fmt.Errorf("managed-fusion-agent-config secret does not contain smtp_config entry")
+		}
+		if err := yaml.Unmarshal(managedFusionDeploymentSMTPConfig, r.smtpConfigData); err != nil {
+			return ctrl.Result{}, err
+		}
+
+		if err := r.reconcileAlertRelabelConfigSecret(); err != nil {
+			return ctrl.Result{}, err
+		}
+		if err := r.reconcilePrometheusKubeRBACConfigMap(); err != nil {
+			return ctrl.Result{}, err
+		}
+		if err := r.reconcilePrometheusService(); err != nil {
+			return ctrl.Result{}, err
+		}
+		if err := r.reconcilePrometheus(); err != nil {
+			return ctrl.Result{}, err
+		}
+		if err := r.reconcileAlertmanager(); err != nil {
+			return ctrl.Result{}, err
+		}
+		if err := r.reconcileAlertmanagerSecret(); err != nil {
+			return ctrl.Result{}, err
+		}
+		if err := r.reconcileAlertmanagerConfig(); err != nil {
+			return ctrl.Result{}, err
+		}
+		if err := r.reconcileK8SMetricsServiceMonitor(); err != nil {
+			return ctrl.Result{}, err
+		}
+		if err := r.reconcileMonitoringResources(); err != nil {
+			return ctrl.Result{}, err
+		}
+		if err := r.reconcilePrometheusProxyNetworkPolicy(); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// TODO Add logic to check if the offerings exists
+func (r *ManagedFusionReconciler) verifyOfferringsDoNotExist() bool {
+	return true
+}
+
+func (r *ManagedFusionReconciler) reconcilePrometheus() error {
+	r.Log.Info("Reconciling Prometheus")
+
+	_, err := ctrl.CreateOrUpdate(r.ctx, r.Client, r.prometheus, func() error {
+		if err := r.own(r.prometheus); err != nil {
+			return err
+		}
+
+		desired := templates.PrometheusTemplate.DeepCopy()
+		utils.AddLabel(r.prometheus, monLabelKey, monLabelValue)
+
+		// use the container image of kube-rbac-proxy that comes in deployer CSV
+		// for prometheus kube-rbac-proxy sidecar
```

Comment:
This comment mention deployer, please update

Reply:
Updated in latest commit